src/tir/schedule/primitive/block_annotate.cc

```diff
@@ -0,0 +1,315 @@
+                                   const Buffer& buffer) {
+    Optional<StmtSRef> defining_site_sref;
+    bool is_alloc;
+    std::tie(defining_site_sref, is_alloc) = GetBufferDefiningSite(block_sref, buffer);
+    if (!defining_site_sref || !is_alloc) {
+      throw NonAllocatedBufferError(mod, buffer);
+    }
+  }
+
+  Array<ObjectRef> LocationsOfInterest() const final { return {}; }
+  IRModule mod() const final { return mod_; }
+
+ private:
+  IRModule mod_;
+  Buffer buffer_;
+};
+
+class StorageAlignInvalidFactorError : public ScheduleError {
+ public:
+  explicit StorageAlignInvalidFactorError(IRModule mod, int factor)
+      : mod_(std::move(mod)), factor_(factor) {}
+
+  String FastErrorString() const final {
+    return "ScheduleError: The input `factor` of storage_align is expected to be a positive "
+           "number.";
+  }
+
+  String DetailRenderTemplate() const final {
+    std::ostringstream os;
+    os << "The input `factor` of storage_align is expected to be a positive number. However, the "
+          "input `factor` is "
+       << factor_ << ", which is out of the expected range.";
+    return os.str();
+  }
+
+  static void Check(const IRModule& mod, int factor) {
+    if (factor <= 0) {
+      throw StorageAlignInvalidFactorError(mod, factor);
+    }
+  }
+
+  Array<ObjectRef> LocationsOfInterest() const final { return {}; }
+  IRModule mod() const final { return mod_; }
+
+ private:
+  IRModule mod_;
+  int factor_;
+};
+
+class StorageAlignInvalidAnnotationError : public ScheduleError {
+ public:
+  explicit StorageAlignInvalidAnnotationError(IRModule mod, Block block)
+      : mod_(std::move(mod)), block_(std::move(block)) {}
+
+  String FastErrorString() const final {
+    return "ScheduleError: The block annotation for storage align is expected to be an array of "
+           "3-integer-tuples (axis, factor, offset).";
+  }
+
+  String DetailRenderTemplate() const final {
+    std::ostringstream os;
+    os << "The block annotation for storage align is expected to be an array of 3-integer-tuples "
+          "(axis, factor, offset). However, the block annotation with key "
+       << attr::buffer_dim_align << " of the block {0} is "
+       << block_->annotations.at(attr::buffer_dim_align) << ", which is unexpected.";
+    return os.str();
+  }
+
+  static Array<Array<Array<Integer>>> CheckAndGetAnnotation(const IRModule& mod,
+                                                            const Block& block) {
+    // Get existing annotation value.
+    auto it = block->annotations.find(attr::buffer_dim_align);
+    if (it != block->annotations.end()) {
+      if (!IsValidAnnotation(block, (*it).second)) {
+        throw StorageAlignInvalidAnnotationError(mod, block);
+      }
+      return Downcast<Array<Array<Array<Integer>>>>((*it).second);
+    }
+
+    // Create new annotation value
+    Array<Array<Array<Integer>>> storage_align_annotation;
+    storage_align_annotation.resize(block->writes.size());
```

Comment:
The problem of doing this is that it could be sort of confusing when we dump the TVM script, so let's keep the array as flat as possible but keep the necessary checks

Reply:
We requires the storage_align annotation array to be the same size as `block->writes` (if it exists), each element is used for each write buffer.  Alternatively, we can also store buffer index in the annotation. What do you think?

Reply:
Yes, I was proposing the lazily store buffer_index as well: https://github.com/apache/tvm/pull/8693#discussion_r686489336